src/main/scala/mesosphere/marathon/core/health/impl/AppHealthCheckActor.scala

```diff
@@ -0,0 +1,112 @@
+
+/**
+  * This actor aggregates the statuses of health checks at the application level
+  * in order to maintain a global healthiness status for each instance.
+  *
+  * @param eventBus The eventStream to publish status changed events to
+  */
+class AppHealthCheckActor(eventBus: EventStream) extends Actor with StrictLogging {
+  /**
+    * Map of health check definitions of all applications
+    */
+  private var healthChecks: Map[ApplicationKey, Set[HealthCheck]] = Map.empty
+
+  /**
+    *  Map of results of all health checks for all applications.
+    *  Results are optional, therefore the global health status of an instance
+    *  is either:
+    *    unknown (if some results are still missing)
+    *    healthy (if all results are known and healthy)
+    *    not healthy (if all results are known and at least one is unhealthy)
+    */
+  private var healthCheckStates: Map[InstanceKey, Map[HealthCheck, Option[Health]]] =
+    Map.empty
+
+
+  private def computeGlobalHealth(instanceHealthResults: Map[HealthCheck, Option[Health]]): Option[Boolean] = {
+    val instanceHealthResultsSet = instanceHealthResults.values.toSet
+    val healthy = instanceHealthResultsSet.forall(x => x.fold(false)(_.alive))
+    val unhealthy = instanceHealthResultsSet.forall(_.nonEmpty) && !healthy
+    (healthy, unhealthy) match {
+      case (true, _) => Some(true)
+      case (_, true) => Some(false)
+      case _ => Option.empty[Boolean]
+    }
+  }
+
+  private def notifyHealthChanged(applicationKey: ApplicationKey, instanceId: Instance.Id,
+    healthiness: Option[Boolean]): Unit = {
+    logger.debug(s"Instance global health status changed to healthiness=$healthiness " +
+      s"for instance appId:$applicationKey instanceId:$instanceId")
+    eventBus.publish(InstanceHealthChanged(
+      instanceId, applicationKey.version, applicationKey.appId, healthiness))
+  }
+
+  private def healthCheckExists(applicationKey: ApplicationKey, healthCheck: HealthCheck): Boolean =
+    healthChecks.contains(applicationKey) && healthChecks(applicationKey).contains(healthCheck)
+
+  override def receive: Receive = {
+    case AddHealthCheck(appKey, healthCheck) =>
+      healthChecks = healthChecks +
+        (appKey -> (healthChecks.getOrElse(appKey, Set.empty) + healthCheck))
+      logger.debug(s"Add health check $healthCheck to instance appId:${appKey.appId} version:${appKey.version}")
+
+    case RemoveHealthCheck(appKey, healthCheck) =>
+      logger.debug(s"Remove health check $healthCheck from instance appId:${appKey.appId} version:${appKey.version}")
+
+      healthChecks = healthChecks +
+        (appKey -> (healthChecks.getOrElse(appKey, Set.empty) - healthCheck))
+
+      healthCheckStates = healthCheckStates.map(kv => {
+        val newHealthChecks = kv._2.filter(x => x._1 != healthCheck)
+        (kv._1, newHealthChecks)
+      })
+
+    case HealthCheckStatusChanged(appKey, healthCheck, health) =>
+      if (healthCheckExists(appKey, healthCheck)) {
```

Comment:
Seems like perhaps we should log a debug message if the health check status changes and we don't have a record of this health check?

Reply:
I will do that.